keps/993-two-phase-admission/README.md

```diff
@@ -0,0 +1,280 @@
+	// allowing other to try, unblock quota) and retry. Can be set only
+	// when Condition status is set to "False"
+	Retry string = "Retry"
+	// The check will not pass in the near future. It is not worth
+	// to retry. Can be set together with Condition status "False".
+	Reject string = "Reject"
+	// The check might pass if there was fewer workloads. Proceed
+	// with any pending workload preemption. Should be set 
+	// when condition status is still "Unknown" 
+	BookQuotaRequired status = "BookQuotaRequired"
+)
+
+// AdmissionCheckSpec defines the desired state of AdmissionCheck.
+type AdmissionCheckSpec struct {
+	// Name of the controller which will actually perform
+	// the checks. This is the name with which controller indentifies with,
+	// not a K8S pod or deployment name. Cannot be empty. 
+	ControllerName string `json:"controllerName"`
+
+	// How long to keep the workload suspended 
+	// after a failed check (after it transitioned to False).
+	// After that the check state goes to "Unknown".
+	// The default is 15 min. 
+	RetryDelayMinutes *int64 `json:"retryDelayMinutes,omitempty"`
+
+	// A reference to the additional parameters for the check. 
+	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`
+
+	// How should the workload book quota (and possibly preempt others) 
+	// before the check pass. The default is NoBooking.
+	// If different modes are used for different AdmissionChecks,
+	// the most restrictive one is used.
+	QuotaBooking *QuotaBookingMode `json:"quotaBooking,omitempty"`
+}
+
+const (
+	// No booking, unless a check explicitely asks for it reporting quota exhausted.
+	NoBooking QuotaBookingMode = "NoBooking"
+	
+	// Book quota (and preempt) immediately after the basic quota checks 
+	// are done, without waiting for the additional checks to cpmplete.
+	BookAfterQuotaCheck QuotaBookingMode = "BookAfterQuotaCheck"
+)
+
+// Points to where the parameters for the checks are. 
+// As clusterqueue are in cluster scope - this should be
+// a dedicated CRD specific for the Controller.
+type AdmissionCheckParametersReference struct {
+	// ApiGroup is the group for the resource being referenced. 
+	APIGroup string `json:"apiGroup"`
+
+	// Kind is the type of the resource being referenced.
+	Kind string `json:"kind"`
+
+	// Name is the name of the resource being referenced.
+	Name string `json:"name"`
+}
+```
+
+For every workload that is put to a ClusterQueue that has AdmissionChecks
+configured Kueue will add:
+
+* “AdmissionPrecheck” set false to Conditions.
+* “<checkName>” set to "Unknown" for each of the AdmissionChecks to AdmissionCheckConditions.
+
+Kueue will perform the very same checks that it does today, 
+before admitting a workload. However, once the basic checks 
+pass AND there are some AdmissionChecks configured, AND the
+workload is not in on-hold retry state from some check, it will:
+
+1. Fill the Admission field in workload, with the desired flavor assignment. 
```

Comment:
This should happen before, along with setting AdmissionPrecheck, otherwise we could lose the assignment on a controller restart.

However, at this point, we should set the Admitted condition to true.

Reply:
Nvm, I got confused with the paragraphs.

The current proposal makes sense.